ide/gototest/src/org/netbeans/api/gototest/TestOppositesLocator.java

```diff
@@ -0,0 +1,194 @@
+
+        Collection<? extends TestLocator> locators = Lookup.getDefault()
+                                                           .lookupAll(TestLocator.class)
+                                                           .stream()
+                                                           .filter(tl -> tl.appliesTo(fo))
+                                                           .collect(Collectors.toList());
+
+        CountDownLatch allDone = new CountDownLatch(locators.size());
+
+        for (TestLocator locator : locators) {
+            if (locator.appliesTo(fo)) {
+                if (locator.asynchronous()) {
+                    locator.findOpposite(fo, caretOffset, new LocationListener() {
+                        @Override
+                        public void foundLocation(FileObject fo, TestLocator.LocationResult location) {
+                            if (location != null) {
+                                FileObject fileObject = location.getFileObject();
+                                if(fileObject == null) {
+                                    String msg = location.getErrorMessage();
+                                    if (msg != null) {
+                                        DialogDisplayer.getDefault().notify(
+                                                new NotifyDescriptor.Message(msg, NotifyDescriptor.INFORMATION_MESSAGE));
+                                    }
+                                } else {
+                                    locationResults.put(location, fileObject.getName());
+                                }
+                            }
+                            allDone.countDown();
+                        }
+                    });
+                } else {
+                    TestLocator.LocationResult opposite = locator.findOpposite(fo, caretOffset);
+
+                    if (opposite != null) {
+                        FileObject fileObject = opposite.getFileObject();
+                        if (fileObject == null) {
+                            String msg = opposite.getErrorMessage();
+                            if (msg != null) {
+                                DialogDisplayer.getDefault().notify(
+                                        new NotifyDescriptor.Message(msg, NotifyDescriptor.INFORMATION_MESSAGE));
+                            }
+                        } else {
+                            locationResults.put(opposite, fileObject.getName());
+                        }
+                    }
+
+                    allDone.countDown();
+                }
+            }
+        }
+
+        try {
+            allDone.await();
+        } catch (InterruptedException ex) {
+            Exceptions.printStackTrace(ex);
+        }
+
+        return locationResults.entrySet()
+                              .stream()
+                              .map(e -> new NamedLocation(e.getKey(),
```

Comment:
Maybe `NamedLocation` is a redundant structure: the `Location` itself carries the `FileObject` used to derive the location's name. 
Or is it intended that `NamedLocation` will report (somehow) more human-readable names in the future ?

Reply:
Maybe sort-of redundant, but the original code was keeping the `Location` and the name separate (as a key and value in a Map), so I didn't think we should change that. And the name is inferred on one place (here), where it can be easily changed if needed, unlike the case where we would second-guess the name.